clippy_lints/src/new_without_default.rs

```diff
@@ -156,16 +224,128 @@ impl<'tcx> LateLintPass<'tcx> for NewWithoutDefault {
-    self_type_snip: &str,
-    generics_sugg: &str,
-    where_clause_sugg: &str,
-) -> String {
-    #[rustfmt::skip]
-    format!(
-"impl{generics_sugg} Default for {self_type_snip}{where_clause_sugg} {{
+// Check if Self is a unit struct, and avoid any kind of suggestions
+// FIXME: this was copied from DefaultConstructedUnitStructs,
+//        and should be refactored into a common function
+fn is_unit_struct(_cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    if let Adt(def, ..) = ty.kind()
+        && def.is_struct()
+        && let var @ VariantDef {
+            ctor: Some((hir::def::CtorKind::Const, _)),
+            ..
+        } = def.non_enum_variant()
+        && !var.is_field_list_non_exhaustive()
+    {
+        true
+    } else {
+        false
+    }
+}
+
+/// Check if a block contains one of these:
+/// - Empty block with an expr (e.g., `{ Self::default() }`)
+/// - One statement (e.g., `{ return Self::default(); }`)
+fn check_block_calls_default(cx: &LateContext<'_>, block: &hir::Block<'_>) -> Option<Span> {
+    if let Some(expr) = block.expr
+        && block.stmts.is_empty()
+        && check_expr_call_default(cx, expr)
+    {
+        // Block only has a trailing expression, e.g. `Self::default()`
+        return None;
+    } else if let [hir::Stmt { kind, .. }] = block.stmts
+        && let hir::StmtKind::Expr(expr) | hir::StmtKind::Semi(expr) = kind
+        && let hir::ExprKind::Ret(Some(ret_expr)) = expr.kind
+        && check_expr_call_default(cx, ret_expr)
+    {
+        // Block has a single statement, e.g. `return Self::default();`
+        return None;
+    }
+
+    // trim first and last character, and trim spaces
+    let mut span = block.span;
+    span = span.with_lo(span.lo() + BytePos::from_usize(1));
+    span = span.with_hi(span.hi() - BytePos::from_usize(1));
+    span = trim_span(cx.sess().source_map(), span);
+
+    Some(span)
+}
+
+/// Check for `Self::default()` call syntax or equivalent
+fn check_expr_call_default(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+    if let hir::ExprKind::Call(callee, &[]) = expr.kind
+        // FIXME: does this include `Self { }` style calls, which is equivalent,
+        //        but not the same as `Self::default()`?
+        // FIXME: what should the whole_call_expr (3rd arg) be?
+        && is_default_equivalent_call(cx, callee, None)
+    {
+        true
+    } else {
+        false
+    }
+}
+
+fn suggest_default_mismatch_new<'tcx>(
+    cx: &LateContext<'tcx>,
+    span: Span,
+    id: rustc_hir::OwnerId,
+    block: &rustc_hir::Block<'_>,
+    self_ty: Ty<'tcx>,
+    impl_self_ty: &rustc_hir::Ty<'_>,
+) {
+    let self_ty_fmt = self_ty.to_string();
+    let self_type_snip = snippet(cx, impl_self_ty.span, &self_ty_fmt);
+    span_lint_hir_and_then(
+        cx,
+        DEFAULT_MISMATCHES_NEW,
+        id.into(),
+        block.span,
+        format!("consider delegating to the auto-derived `Default` for `{self_type_snip}`"),
```

Comment:
You are supposed to explain the problem here, not give the solution. Right now, there seems to be no indication of what is wrong in the `stderr` file, just a suggestion to delegate. Nowhere does the notion of "mismatch" appear in the output. This text should probably appear as a `diag.help()` message.

Also, is the suggestion accurate when `new()` and `default()` don't match? Shouldn't it suggest to either be consistent and delegate, or (as discussed in the PR comments) rename the `new()` associated function to something more specific?

Should the case where `new()` and `default()` return the same value be considered a mismatch? Or just a redundancy, in which case the lint may be misnamed.